src/runtime/minrpc/minrpc_logger.cc

```diff
@@ -0,0 +1,288 @@
+        LogString(" } ");
+      }
+      break;
+    }
+    default: {
+      LogString("ERROR-kUnknownTypeCode)");
+      break;
+    }
+  }
+  LogString("; ");
+}
+
+void Logger::OutputLog() {
+  LOG(INFO) << os_.str();
+  os_.str(std::string());
+}
+
+void MinRPCReturnsWithLog::ReturnVoid() {
+  next_->ReturnVoid();
+  logger_.LogString("-> ReturnVoid");
+  logger_.OutputLog();
+}
+
+void MinRPCReturnsWithLog::ReturnHandle(void* handle) {
+  next_->ReturnHandle(handle);
+  if (code_ == RPCCode::kGetGlobalFunc) {
+    RegisterHandleName(handle);
+  }
+  logger_.LogVal<void*>("-> ReturnHandle: ", handle);
+  logger_.OutputLog();
+}
+
+void MinRPCReturnsWithLog::ReturnException(const char* msg) {
+  next_->ReturnException(msg);
+  logger_.LogString("-> Exception: ");
+  logger_.LogString(msg);
+  logger_.OutputLog();
+}
+
+void MinRPCReturnsWithLog::ReturnPackedSeq(const TVMValue* arg_values, const int* type_codes,
+                                           int num_args) {
+  next_->ReturnPackedSeq(arg_values, type_codes, num_args);
+  ProcessValues(&arg_values, &type_codes, &num_args);
+  logger_.OutputLog();
+}
+
+void MinRPCReturnsWithLog::ReturnCopyAck(uint64_t* num_bytes, uint8_t** data_ptr) {
+  next_->ReturnCopyAck(num_bytes, data_ptr);
+  logger_.LogVal<uint64_t>("-> CopyAck: ", *num_bytes);
+  logger_.LogVal<void*>(", ", reinterpret_cast<void*>(*data_ptr));
+  logger_.OutputLog();
+}
+
+void MinRPCReturnsWithLog::ReturnLastTVMError() {
+  const char* err = TVMGetLastError();
+  ReturnException(err);
+}
+
+void MinRPCReturnsWithLog::ThrowError(RPCServerStatus code, RPCCode info) {
+  next_->ThrowError(code, info);
+  logger_.LogString("-> ERROR");
+  logger_.OutputLog();
+}
+
+void MinRPCReturnsWithLog::ProcessValues(const TVMValue** values, const int** tcodes,
+                                         int* num_args) {
+  if (*tcodes != nullptr) {
+    logger_.LogString("-> [");
+    for (int i = 0; i < *num_args; ++i) {
+      logger_.LogTVMValue((*tcodes)[i], (*values)[i]);
+
+      if ((*tcodes)[i] == kTVMOpaqueHandle) {
+        RegisterHandleName((*values)[i].v_handle);
+      }
+    }
+    logger_.LogString("]");
+  }
+}
+
+void MinRPCReturnsWithLog::ResetCurrHandleName(RPCCode code) {
+  code_ = code;
+  logger_.LogString(RPCCodeToString(code));
+  logger_.LogString(", ");
+  curr_handle_name_.clear();
+}
+
+void MinRPCReturnsWithLog::UpdateCurrHandleName(const char* name) {
+  if (curr_handle_name_.length() != 0) {
+    curr_handle_name_.append("::");
+  }
+  curr_handle_name_.append(name);
+}
+
+void MinRPCReturnsWithLog::GetHandleName(void* handle) {
+  if (array_tracker_.find(handle) != array_tracker_.end()) {
+    curr_handle_name_.append(array_tracker_[handle]);
+    logger_.LogHandleName(curr_handle_name_);
+  }
+}
+
+void MinRPCReturnsWithLog::ReleaseHandleName(void* handle) {
+  if (array_tracker_.find(handle) != array_tracker_.end()) {
+    logger_.LogHandleName(array_tracker_[handle]);
```

Comment:
wondering if we need to Log at the same time we register?

Reply:
ReleaseHandleName is called when we receive the FreeHandle SysCall. In that case, I check if the name is registered, print it if it is, and remove it from the map of registered names.